### PR TITLE
Fix bug with CoT early stopping and LLama2 tokenizer

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?

LLaMa2 tokenizer didn't work with early stopping because it inserted some extra token ([28971](https://github.com/huggingface/transformers/issues/26273)) before \n\n. This caused the stop sequence to appear longer than it really was, which resulted in the stopping criteria looking back too far into the input and causing generation to stop as soon as it begins. With the new implementation we remove any superfluous tokens in the stopping criteria (i.e. those tokens that add nothing to the decoded result) I also reduced the amount of extra lookback we do to make it so we don't ever look back farther than we should and stop as soon as generation begins.

Testing: `mpt-eval-NzfmSy`

```
| Category   | Benchmark   | Subtask   |   Accuracy | Number few shot   | Model                    |
|:-----------|:------------|:----------|-----------:|:------------------|:-------------------------|
|            | gsm8k       |           |   0.141774 | 0-shot            | meta-llama/Llama-2-7b-hf |
```

Results replicate LLama2 paper (14.6 for 7b)
# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: tests/datasets/test_in_context_learning_datasets.py [hunk 1]
```diff
@@ -83,6 +83,24 @@ def test_stop_sequences_criteria(tiny_gpt2_tokenizer):
     assert eos_criteria(input_ids, None)
 
 
+def test_stop_sequences_criteria_sentencepiece():
+    pytest.importorskip('datasets')
+
+    tokenizer = AutoTokenizer.from_pretrained('huggyllama/llama-7b', use_fast=False)
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:89]: please use the llama tokenizer fixture not download this one

File: composer/datasets/utils.py [hunk 2]
```diff
@@ -186,20 +186,22 @@ def __init__(
...
+
+            # we look back for 1 more token than it takes to encode our stop sequence
             # because tokenizers suck, and a model might generate `['\n', '\n']` but our `sequence` is `['\n\n']`
             # and we don't want to mistakenly not stop a generation because our
             # (string) stop sequence was output in a different tokenization
 
-            # NOTE: there is a minor danger that this will end up looking back 2 tokens into the past, into the inputs to the model,
-            # and stopping generation immediately as a result. With only 2 extra tokens of lookback, this risk is minimized
-            self.stop_sequence_id_len = len(self.stop_sequence_ids) + 2
+            self.stop_sequence_id_len = len(self.stop_sequence_ids) + 1
```
Comment (dakinggg) [inline composer/datasets/utils.py:199]: why did you change this to +1

**Comment by dakinggg** [comment]
Can you please include some evidence this resolves the issue in the pr description?

File: composer/datasets/utils.py (hunk 2 above)
Comment (bmosaicml) [reply composer/datasets/utils.py:199]: since the model generates 1 token before the first call to the stopping criteria, this ensures we will never stop too early for any of the stopping criteria we use in practice (since they're all multi-token).

File: tests/datasets/test_in_context_learning_datasets.py [hunk 3]
```diff
@@ -220,13 +220,31 @@ def test_stop_sequences_criteria(tiny_gpt2_tokenizer):
     seq2 = tiny_gpt2_tokenizer('Dogs are furry\n\n')['input_ids']
     seq1 = [50257] * (len(seq2) - len(seq1)) + seq1
     input_ids = torch.tensor([seq1, seq2])
-    assert not eos_criteria(input_ids, None)
+    assert not eos_criteria(input_ids.type(torch.long), None)  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:223]: instead of this, you just want to construct `LongTensor` on the line above

File: tests/datasets/test_in_context_learning_datasets.py [hunk 4]
```diff
@@ -220,13 +220,31 @@ def test_stop_sequences_criteria(tiny_gpt2_tokenizer):
...
 
     eos_criteria = MultiTokenEOSCriteria('\n\n', tiny_gpt2_tokenizer, 2)
     seq1 = tiny_gpt2_tokenizer('Dogs are furry\n\n')['input_ids']
     seq2 = tiny_gpt2_tokenizer('Dogs are furry\n\n')['input_ids']
     input_ids = torch.tensor([seq1, seq2])
-    assert eos_criteria(input_ids, None)
+    assert eos_criteria(input_ids.type(torch.long), None)  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:229]: same

File: tests/datasets/test_in_context_learning_datasets.py [hunk 5]
```diff
@@ -220,13 +220,31 @@ def test_stop_sequences_criteria(tiny_gpt2_tokenizer):
...
+
+
+def test_stop_sequences_criteria_sentencepiece(tiny_llama_tokenizer):
+    pytest.importorskip('datasets')
+
+    tokenizer = tiny_llama_tokenizer
+    eos_criteria = MultiTokenEOSCriteria('\n\n', tokenizer, 2)
+    seq1 = tokenizer('\n\nDogs')['input_ids']  # check to make sure starting with the start sequence doesnt break it
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:237]:
```suggestion
    seq1 = tokenizer('\n\nDogs')['input_ids']  # check to make sure starting with the stop sequence doesnt break it
```
?

File: tests/datasets/test_in_context_learning_datasets.py [hunk 6]
```diff
@@ -220,13 +220,31 @@ def test_stop_sequences_criteria(tiny_gpt2_tokenizer):
...
+    seq2 = tokenizer('Dogs are furry\n\n')['input_ids']
+    seq1 = [50257] * (len(seq2) - len(seq1)) + seq1
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:239]: what is 50257 supposed to do here? It is outside the vocab of llama

**Review by dakinggg — APPROVED** [review]
